MennaMohie/assignment1/Q2.cpp

```diff
@@ -27,37 +26,88 @@ class SinglyLinkedList
     int Size;
 
 public:
-    SinglyLinkedList():head(NULL),tail(NULL),Size(0) {}
+    SinglyLinkedList():head(nullptr),tail(nullptr),Size(0) {}
+    SinglyLinkedList(const SinglyLinkedList<T> &SLL)
+    {
+        this->Size = SLL.Size;
+
+        if(SLL.head==nullptr)
+        {
+            head=nullptr;
+            tail=nullptr;
+        }
+        else
+        {
+            head=new Node<T>(SLL.head->value);
+            Node<T> *this_temp_ptr = head;
+            Node<T> *other_list_ptr = SLL.head->next;
+            while(other_list_ptr!=nullptr)
+            {
+                this_temp_ptr->next = new Node<T>(other_list_ptr->value);
+                other_list_ptr=other_list_ptr->next;
+                this_temp_ptr=this_temp_ptr->next;
+            }
+            this->tail = this_temp_ptr;
 
-    void DisplayKthToLastNode(int k)
+        }
+    }
+    SinglyLinkedList<T>& operator = (const SinglyLinkedList<T> &SLL)
     {
-        assert(k>=0 && k<Size);
-        int pos=Size-k-1;
-        DisplayAt(pos);
+        this->Size = SLL.Size;
 
+        if(SLL.head==nullptr)
+        {
+            head=nullptr;
+            tail=nullptr;
+        }
+        else
+        {
+            head=new Node<T>(SLL.head->value);
+            Node<T> *this_temp_ptr = head;
+            Node<T> *other_list_ptr = SLL.head->next;
+            while(other_list_ptr!=nullptr)
+            {
+                this_temp_ptr->next = new Node<T>(other_list_ptr->value);
+                other_list_ptr=other_list_ptr->next;
+                this_temp_ptr=this_temp_ptr->next;
+            }
+            this->tail = this_temp_ptr;
+        }
+        return *this;
     }
 
-    void DisplayAt(int pos)
+    T GetKthToLastNode(int k)
```

Comment:
This declaration leaves you no room to return a error in case k is out of bounds, either:
* return a success bool, and have a `T *` parameter to put the node. (preferred)
* return a `const T *`, using nullptr as error/out of bounds.
* Throw an exception (not recommneded)